File: hackerrank/problem solving/algorithms/2 - implementation/36 - repeated string/repeated_string.go
// https://www.hackerrank.com/challenges/repeated-string/problem?isFullScreen=true

package main

import (
    "fmt"
    "strings"
)

func main() {
    var str string
    var nChars int
    _, _ = fmt.Scan(&str, &nChars)
    const letter = "a"
    fmt.Println(countInRepeatedString(str, letter, nChars))
}

// n: length of str
// T: O(n)
// S: O(1) extra space
func countInRepeatedString(str, letter string, nChars int) int {
    fullRepeats := nChars / len(str)
    nSubstr := nChars % len(str)

    count := strings.Count(str, letter)
    count *= fullRepeats
    return count + strings.Count(str[:nSubstr], letter)
}

File: hackerrank/problem solving/algorithms/2 - implementation/36 - repeated string/repeated_string.cs
// https://www.hackerrank.com/challenges/repeated-string/problem?isFullScreen=true

string str = Console.ReadLine()!;
long nChars = long.Parse(Console.ReadLine()!);
const char letter = 'a';
Console.WriteLine(CountInRepeatedString(str, letter, nChars));

// n: length of str
// T: O(n)
// S: O(1) extra space
long CountInRepeatedString(string str, char letter, long nChars)
{
    long nSubstr;
    long fullRepeats = Math.DivRem(nChars, str.Length, out nSubstr);
    var substr = str.AsSpan(0, (int)nSubstr);

    long count = str.Count(c => c == letter);
    count *= fullRepeats;
    return count + substr.Count(letter);
}
